File: cloudflare-backend/src/routes/reports.ts
// src/routes/reports.ts
import { Hono } from 'hono';
import { AppEnv } from '../middleware/auth';
import { ReportSummary } from '../types';

const reports = new Hono<AppEnv>();

reports.get('/meals', async (c) => {
  const row = await c.env.DB.prepare('SELECT COUNT(*) as count FROM meal_records').first<{ count: number }>();
  const summary: ReportSummary = {
    total_records: row?.count || 0,
    generated_at: new Date().toISOString(),
  };
  return c.json(summary);
});

reports.get('/transactions', async (c) => {
  const row = await c.env.DB.prepare('SELECT COUNT(*) as count FROM transactions').first<{ count: number }>();
  const summary: ReportSummary = {
    total_records: row?.count || 0,
    generated_at: new Date().toISOString(),
  };
  return c.json(summary);
});

reports.get('/users', async (c) => {
  const row = await c.env.DB.prepare('SELECT COUNT(*) as count FROM users').first<{ count: number }>();
  const summary: ReportSummary = {
    total_records: row?.count || 0,
    generated_at: new Date().toISOString(),
  };
  return c.json(summary);
});

reports.get('/export', async (c) => {
  const meals = await c.env.DB.prepare('SELECT COUNT(*) as count FROM meal_records').first<{ count: number }>();
  const txns = await c.env.DB.prepare('SELECT COUNT(*) as count FROM transactions').first<{ count: number }>();
  const users = await c.env.DB.prepare('SELECT COUNT(*) as count FROM users').first<{ count: number }>();

  const dateStr = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 16);
  const filename = `mianu_export_${dateStr}.csv`;

  return c.json({
    status: 'ready',
    path: `/downloads/${filename} (${users?.count || 0} delegates, ${meals?.count || 0} meals, ${txns?.count || 0} txns)`,
  });
});

export default reports;

File: backend/app/routers/reports.py
from datetime import datetime

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.models.models import MealRecord, TransactionRecord, User
from app.schemas import ReportSummary

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/meals", response_model=ReportSummary)
def meals_report(db: Session = Depends(get_db)) -> ReportSummary:
    count = db.query(MealRecord).count()
    return ReportSummary(total_records=count, generated_at=datetime.utcnow())


@router.get("/transactions", response_model=ReportSummary)
def transactions_report(db: Session = Depends(get_db)) -> ReportSummary:
    count = db.query(TransactionRecord).count()
    return ReportSummary(total_records=count, generated_at=datetime.utcnow())


@router.get("/users", response_model=ReportSummary)
def users_report(db: Session = Depends(get_db)) -> ReportSummary:
    count = db.query(User).count()
    return ReportSummary(total_records=count, generated_at=datetime.utcnow())


@router.get("/export")
def export_report(db: Session = Depends(get_db)) -> dict[str, str]:
    meals_count = db.query(MealRecord).count()
    txns_count = db.query(TransactionRecord).count()
    users_count = db.query(User).count()
    timestamp = datetime.utcnow().strftime("%Y-%m-%d_%H-%M")
    filename = f"mianu_export_{timestamp}.csv"
    return {
        "status": "ready",
        "path": f"/downloads/{filename} ({users_count} delegates, {meals_count} meals, {txns_count} txns)",
    }
